src/NUnitFramework/framework/Constraints/ComparisonConstraint.cs

```diff
@@ -77,6 +77,37 @@ public override ConstraintResult ApplyTo<TActual>(TActual actual)
             return new ConstraintResult(this, actual, PerformComparison(_comparer, actual, _expected, _tolerance));
         }
 
+        /// <summary>
+        /// The Description of what this constraint tests, for
+        /// use in messages and in the ConstraintResult.
+        /// </summary>
+        public override string Description
+        {
+            get
+            {
+                System.Text.StringBuilder sb = new System.Text.StringBuilder(DisplayMessageBase);
+                sb.Append(MsgUtils.FormatValue(_expected));
+                
+                if (_tolerance != null && !_tolerance.IsUnsetOrDefault)
+                {
+                    sb.Append(" within ");
+                    sb.Append(MsgUtils.FormatValue(_tolerance.Amount));
+
+                    if (_tolerance.Mode == ToleranceMode.Percent)
+                    {
+                        sb.Append(" percent");
+                    }
+                }
+                
+                return sb.ToString();
+            }
+        }
+
+        /// <summary>
+        /// Protected field overriden by derived class to provide basis for Description construction
+        /// </summary>
+        protected abstract string DisplayMessageBase { get; }
```

Comment:
How flexible does it have to be? It's the base class for four of our classes... the four comparison constraints. It's not a class we offer for extension as far as I can see.

Reply:
How will folks know they're not intended to subclass `ComparisonConstraint`?

Typically that design decision is implemented using `internal abstract` (or better, `private protected abstract` in C# 7.0) so that the only assemblies that can extend it are assemblies with internal access to this one.

Reply:
This is a general discussion you and I have been having for a few years now and it continues to infect discussions of particular changes, where the person performing the work if the change just needs to be told what to do.

We won't resolve our different views of what constitutes an API here - probably ever - so let's not go there on this issue. 

Instead, why not just have this developer change the signature of the base class?

Reply:
"Infect" is not a diplomatic word. 😆

By changing the signature, you mean asking for `internal virtual` instead of `protected abstract`? Okay, but it's not my place (or desire) to make unilateral decisions. I would like to solicit thoughts like yours from the team. Definitions of APIs aside, is there something which differentiates ComparisonConstraint from Constraint in such a way that

- we expect no users to have inherited from `ComparisonConstraint`
- or, we want users who inherit to be required by the compiler to figure out what we did and change their programs when they upgrade?

Decisions can certainly be made by default but to me it's worth being sure that we like the outcome of this decision.

Reply:
@jnm2 You should see the words I rejected. :smiley:

Seriously when we (you and I) have these broader discussions, the rest of the team loses interest in the thread. That's why I call it an infection. So let's just not do that.

As you well know, I no longer vote or take part in discussions of where nunit will go in the future, because I'm not a part of that future. I do try to strike a balance by giving you all some info from the past where it seems necessary from my perspective. It may be that that's not working, in which case I'll stop.

Reply:
And yet I specifically asked _not_ to have a broader discussion about APIs. Can we focus on this single, direct outcome of `protected abstract`? Like I asked above: 1) Will it affect anyone? 2) If so, do we like the effect, or should we use a less obvious alternative like `internal` or `virtual`?

@nunit/framework-team If your preference is that I skip asking this kind of end user cause-effect question and simply request our (awesome) contributor to take what I think is the safer route, I can sure do that.

Reply:
Thank you for the detailed discussion. 
It was my bad not paying attention to the class being public and possibility (even slightest) of it being used in someone else's code. I guess @jnm2 has the point here, and it will be better solution to make it virtual in order to not break things up.

Reply:
If you mean defining the `DescriptionBase` property as protected abstract, it seems fine to me as he has done it. You seemed to be critiquing that change because that's where your comment took off from.

If you are suggesting changing the signature of `ComparisonConstraint`, that's what seems to me to be out of scope here and part of a broader discussion of how we handle classes that are intended only to be inherited from by other classes of ours. That's the broader discussion I thought you were trying to initiate, even while saying you didn't want to initiate it.

Reply:
@CharliePoole Yes, I'm only talking about `DescriptionBase`. I'm not sure what "changing the signature of ComparisonConstraint" means; did it sound like I was suggesting making the whole class internal? That's wouldn't compile because public classes can only inherit from other public classes. My bad if that wasn't clear.

@madetara There's two things I saw. One was that custom inheritors would have to implement that method even though they're currently overriding `Description` if they are setting the description at all.  Different team members have expressed a desire to minimize required changes like this unless the custom implementation is already buggy in the previous release.

The other thing is that we can't remove `DescriptionBase` in the future, even if we find out down the road that we need something else and `DescriptionBase` becomes vestigial. This isn't a huge problem, just worth a passing thought.

A tactic that doesn't require derived classes to change their understanding of how the base class works might be something like adding an internal or protected (sealed) member `string DefaultDescription(string comparisonText)` to ComparisonConstraint. Derived classes could then override Description with `=> DefaultDescription("greater than")`, etc.

Reply:
> @nunit/framework-team If your preference is that I skip asking this kind of end user cause-effect question and simply request our (awesome) contributor to take what I think is the safer route, I can sure do that. So far I've asked for your input because I like hearing from you and I'm just one member here.

Personally, I value that you catch these kinds of issues (I think that I would have overlooked it). I'm happy with you requesting the contributor to do the changes, then we can always discuss if we disagree. At least for me I think that I would look just as closely on a normal PR review comment as on the original comment https://github.com/nunit/nunit/pull/3044#discussion_r223183731, but of course it also depends on scale/complexity of the changes (the larger the more need for discussion and consensus).

Ps. I like the last variant proposed (which is also implemented now).

Reply:
@mikkelbu Okay, sounds like a plan! Thanks!